src/hotspot/share/opto/compile.cpp

```diff
@@ -3298,6 +3298,24 @@ void Compile::final_graph_reshaping_main_switch(Node* n, Final_Reshape_Counts& f
   case Op_Opaque1:              // Remove Opaque Nodes before matching
     n->subsume_by(n->in(1), this);
     break;
+  case Op_CallLeafPure: {
+    if (!Matcher::match_rule_supported(Op_CallLeafPure)) {
+      CallNode* call = n->as_Call();
+      CallNode* new_call = new CallLeafNode(call->tf(), call->entry_point(),
+                                            call->_name, TypeRawPtr::BOTTOM);
+      new_call->init_req(TypeFunc::Control, call->in(TypeFunc::Control));
+      new_call->init_req(TypeFunc::I_O, C->top());
+      new_call->init_req(TypeFunc::Memory, C->top());
+      new_call->init_req(TypeFunc::ReturnAdr, C->top());
+      new_call->init_req(TypeFunc::FramePtr, C->top());
+      for (unsigned int i = 0; i < call->tf()->domain()->cnt() - TypeFunc::Parms; i++) {
+        new_call->init_req(TypeFunc::Parms + i, call->in(TypeFunc::Parms + i));
+      }
```

Comment:
The `TypeFunc::Parms` offsets are a bit confusing / unnecessary. Why not:
```suggestion
      // Copy all <something>.
      for (unsigned int i = TypeFunc::Parms; i < call->tf()->domain()->cnt(); i++) {
        new_call->init_req(i, call->in(i));
      }
```

Reply:
Fine with me! Looks perfectly identical to me. Comment seems superfluous to me: the code is very simple, let's not distract the reader.